### ISSUE TITLE
Custom RetryPolicy is being evaluated twice per SLR defer

### ISSUE THREAD

**Issue by indualagarsamy** [Particular/NServiceBus#3443]
Raised by @nfdias originally here: https://github.com/Particular/docs.particular.net/issues/1124

Problem:
In NSB v5.2.14, SecondLevelRetries.CustomRetryPolicy gets executed more number of times than it should.

In the following example, the CustomPolicy should've got executed once, delayed by 10 seconds and then executed one more time and the message should have been forwarded to the error queue. However, the method is being invoked 3 times!

Repro:
```c#
    using System;
    using NServiceBus;
    using NServiceBus.Logging;

    class Program
    {
        static void Main(string[] args)
        {
            LogManager.Use<DefaultFactory>().Level(LogLevel.Info);
            BusConfiguration busConfiguration = new BusConfiguration();
            busConfiguration.EndpointName("TestCustomPolicy");
            busConfiguration.UseSerialization<JsonSerializer>();
            busConfiguration.UsePersistence<InMemoryPersistence>();
            busConfiguration.EnableInstallers();
            busConfiguration.SecondLevelRetries().CustomRetryPolicy(new NServiceBusSecondLevelHandling().RetryPolicy);


            using (IBus bus = Bus.Create(busConfiguration).Start())
            {
                while (Console.ReadLine() != null)
                {
                    bus.SendLocal(new DoSomething());
                }
            }
        }
    }

    public class NServiceBusSecondLevelHandling
    {
        public TimeSpan RetryPolicy(TransportMessage transportMessage)
        {
            Console.WriteLine("Inside custom retry policy");
            string value;

            int retries = transportMessage.Headers.TryGetValue(Headers.Retries, out value) ? int.Parse(value) : 0;
            
            if (retries >= 1)
            {
                return TimeSpan.MinValue;
            }

            return TimeSpan.FromSeconds(10);
        }
    }

  

    public class DoSomething : ICommand
    {
    }

    class DoSomethingHandler : IHandleMessages<DoSomething>
    {
        public void Handle(DoSomething message)
        {
            Console.WriteLine("received message");
            throw new Exception("Something Bad Happened");
        }
    }
```

**Comment by andreasohlund** [comment]
Since this cause retries to happen for a longer than requested by users this will risk breaking their message SLA's, we need to patch this.

**Comment by boblangley** [comment]
The SLR retries are not being duplicated, just evaluation of the function. The RetryPolicy is called once in the [FaultManager](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Faults/Forwarder/FaultManager.cs#L79) before the message is sent to the retry queue and then again in the [SecondLevelRetryProcessor](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/SecondLevelRetries/SecondLevelRetriesProcessor.cs#L40) in the satellite.

I have created a [repro](https://github.com/BobAtParticular/NSB.Issue3443) and confirmed this behavior is not present in 4.7.4.

Since this breaks SimVer we should patch this.

Pull https://github.com/Particular/NServiceBus/pull/2509 introduced this second call with this description:

> This allows us to keep the notification error stream as typed exceptions.
> Fixed issue with notifications acceptance test would sometimes miss events.
> Reusing default SLR settings for both config section and FaultManager.
> Added validation rules to SecondLevelRetriesConfig section.

Perhaps @johnsimons or @andreasohlund can give us some additional context.

I have submitted Pull request #3453 with a failing acceptance test. No change to the behavior has been implemented.

I see 2 ways forward:

1. Revert the call in FaultManager.
2. Remove the call in SecondLevelRetriesProcessor,using a value embedded in the message header instead.

**Comment by boblangley** [comment]
re # 2 We would probably want to do the call in SecondLevelRetriesProcessor if the header value isn't present to maintain backwards compatibility as per [the comment in SecondLevelRetriesProcessor](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/SecondLevelRetries/SecondLevelRetriesProcessor.cs#L37).

**Comment by andreasohlund** [comment]
We need to make a call on this one since #3468 need to go out as well. Do we include this in 5.2.15 or bump it to 5.2.16?

**Comment by andreasohlund** [comment]
@indualagarsamy I know that you see it differently but I'm no longer convinced that this is a patch since we actually don't perform the actual retry more than requested. 

Yes we call the user "func" a few extra times so if they do some crazy stuff in there it might be some issues but I honestly don't see more as a minor flaw.

How about:

1. We add a doco warning clarifying this
2. Post to the group to make users more aware
3. Fix this for v6

Thoughts?

**Comment by SimonCropp** [comment]
@andreasohlund crazy stuff? Isnt it likely that a user would be doing some kind of counting in there?

**Comment by indualagarsamy** [comment]
The way I see it is, we broke this behavior in a patch.
Up until 4.7.4, we used to execute it x times, post that version, we execute it x+1 times? 
What if the user had some code in that func that's getting executed more times?

**Comment by SimonCropp** [comment]
@indualagarsamy first

**Comment by boblangley** [comment]
@andreasohlund We are SemVer. This bug changed behavior, whether we intended to or not. It should be patched just because of this.

I started a doco [pull request](https://github.com/Particular/docs.particular.net/pull/1191) for v6.

**Comment by andreasohlund** [comment]
> Isn't it likely that a user would be doing some kind of counting in there?

True, that counter would be skewed.

>This bug changed behavior, whether we intended to or not
There is a grey zone here, its partly up to us to define what we behavior is intended and what is a "bug". This is obviously not a bug so I agree.

@SimonCropp @indualagarsamy @boblangley thanks for helping me understand this better. I'm :+1: for patching this, please sync with @ramonsmits regarding the patch he's working on

**Comment by danielmarbach** [comment]
I agree with @boblangley This is a behavior breaking change which violates the principles of least surprise ;) I as a customer would assume that this delegate only gets invoked once per retry.

**Comment by danielmarbach** [comment]
While talking to @udidahan in today's platform sync I realized I got trapped by the whole semantic version and behavior breaking change discussion. So I'm changing my mind. In the discussion in this thread we haven't come up with a scenario how this would lead to message loss. 

Looking out our [backporting recommendations(https://github.com/Particular/PlatformDevelopment/blob/eccbbf5a9b8779819fef7142626941c3ff965bd1/Guidelines/Patching.md#recommendations-for-backporting) the only thing I could find is "Incorrect business behavior". All the other scenarios are not fulfilled. 

So :-1: for backporting.

**Comment by bording** [comment]
If the custom RetryPolicy was doing some sort of internal message tracking where they were setting up a variable retry timespan based on number of times it went through SLR, then calling the RetryPolicy twice for a single run through the retry queue could definitely end up with incorrect/unexpected behavior here.

If "Incorrect business behavior" is a valid reason for backporting a fix, wouldn't this be enough to qualify?

**Comment by boblangley** [comment]
I have been on both sides of this issue.

From a "triage" point of view, especially considering _known_ customer impact, this is not a critical issue. Even in the case raised by the customer, having it called twice does not change the behavior because he is only using the inputs given (TransportMessage) for state. This was the implicit expectation we had for a customer implemented Func. So informing the customer and documenting the behavior could be an acceptable resolution.

Except we say we are SemVer. That was the reason I changed my initial tragectory to this issue from _document only_ to _patch_. We changed the behavior, we should patch it back.

No matter what, I do not think there is any reason to rush out a patch.

So it is a question of whether we should patch to remain compliant (doing what we said we were going to do) or be reasonable considering the low scope and impact of this issue.

I personally feel that if we committed to SemVer we should go through with the patch. If I am mistaken in that assumption, then we should just document, inform the customer that raised the issue, and move on.

**Comment by johnsimons** [comment]
Here is another perspective, this breaking behavior change was introduced in v 4.7.3, which was released more then 1 year ago, it could be that any user using a higher version is aware of the new behavior and maybe even taking advantage of it.
So by us "fixing" it we could be possible breaking these users. This is the reason majority of times MS does not fix broken behavior in previous versions, because there is always the other side of the coin :wink:

**Comment by boblangley** [comment]
Just as a reminder, in my [reproduction](https://github.com/BobAtParticular/NSB.Issue3443) I have confirmed this only occurs in 5.1 and 5.2, but your point is still taken @johnsimons :smile:

**Comment by udidahan** [comment]
If it is low impact, should it not be labeled as such?

**Comment by adamralph** [comment]
> "CustomRetryPolicy code executed multiple times than it should."

Something wrong in that sentence. Can we rephrase?

**Comment by boblangley** [comment]
Since we moved this to the back log, does this mean we will eventually patch 5.1/5.2 or should we mark this as will not fix?

**Comment by danielmarbach** [comment]
@ramonsmits is preparing a new patch release. Maybe get in touch?

On Friday, February 26, 2016, Bob Langley <notifications@github.com> wrote:

> Since we moved this to the back log, does this mean we will eventually
> patch 5.1/5.2 or should we mark this as will not fix?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/issues/3443#issuecomment-189511767>
> .
>


-- 
*Don't miss the Async/Await Webinar Series*
*• Learn more <http://go.particular.net/l/27442/2016-02-15/681bjl> →*

**Comment by udidahan** [comment]
Things in the backlog are not supposed to be worked on.

**Comment by boblangley** [comment]
**TLDR: I propose we clarify intended use and close this as `Withdrawn: Don't Fix` since no customers have been demonstrated to be adversely impacted by this minor behavior change.**

I think the only argument *for* patching this is because of the change in behavior may break SemVer. The public API for NServiceBus was not changed, and we did not have explicit test coverage that established what the "correct" behavior is supposed to be. However I do believe we can make a case that there was an intended use, which was for the user to use the properties of the TransportMessage, such as Retry count, to determine the SLR policy.

This is what we do in our code and it is [how our documentation frames expected usage](http://docs.particular.net/nservicebus/errors/automatic-retries#second-level-retries-custom-retry-policy-error-headers-helper). We have no evidence any users have been impacted by this behavior.

I do not think we should patch this. Patching this would mean either removing performance improvements users on the 5.1/5.2 branches are currently enjoying or introducing additional headers that will only serve to work this specific issue. Neither of these solutions have much, if any, value to our customers.

We can make our documentation for this feature more explicit to help any potential future users that may become confused by this behavior or implement this in a way that is not recommended.

**Comment by andreasohlund** [comment]
Since it's fixed on v6, lets assign to the v6 milestone and close?

**Comment by adamralph** [comment]
@andreasohlund :+1: fixed in v6, no backport.

**Comment by timbussmann** [comment]
closing this and assigning to v6 as suggested by @andreasohlund
